crates/services/gas_price_service/src/fuel_gas_price_updater/da_source_adapter/service.rs

```diff
@@ -13,39 +11,93 @@ use std::{
     time::Duration,
 };
 use tokio::{
-    sync::mpsc::Sender,
+    sync::broadcast::Sender,
     time::{
         interval,
         Interval,
     },
 };
 
 pub use anyhow::Result;
+use tokio::sync::broadcast::Receiver;
+
+pub struct DaBlockCostsProvider<Source: DaBlockCostsSource + 'static> {
+    handle: ServiceRunner<DaBlockCostsService<Source>>,
+    subscription: Receiver<DaBlockCosts>,
+}
+
+#[async_trait::async_trait]
+pub trait DaBlockCostsProviderPort {
+    async fn start_and_await(&self) -> Result<()>;
+    async fn stop_and_await(&self) -> Result<()>;
+    async fn recv(&mut self) -> Result<DaBlockCosts>;
+    fn try_recv(&mut self) -> Result<DaBlockCosts>;
+}
+
+#[async_trait::async_trait]
+impl<Source> DaBlockCostsProviderPort for DaBlockCostsProvider<Source>
+where
+    Source: DaBlockCostsSource,
+{
+    async fn start_and_await(&self) -> Result<()> {
+        self.handle.start_and_await().await?;
+        Ok(())
+    }
+
+    async fn stop_and_await(&self) -> Result<()> {
+        self.handle.stop_and_await().await?;
+        Ok(())
+    }
+
+    async fn recv(&mut self) -> Result<DaBlockCosts> {
+        self.subscription.recv().await.map_err(Into::into)
+    }
+
+    fn try_recv(&mut self) -> Result<DaBlockCosts> {
+        self.subscription.try_recv().map_err(Into::into)
+    }
+}
+
+#[derive(Clone)]
+struct DaBlockCostsServiceSharedState {
+    sender: Sender<DaBlockCosts>,
+}
+
+impl DaBlockCostsServiceSharedState {
+    pub fn subscribe(&self) -> Receiver<DaBlockCosts> {
+        self.sender.subscribe()
+    }
+
+    fn send(&self, da_block_costs: DaBlockCosts) -> Result<()> {
+        self.sender.send(da_block_costs)?;
+        Ok(())
+    }
```

Comment:
[nit] You have access to private fields of this structure, so maybe we don't need this helper function

Reply:
addressed in 69651a53a